### ISSUE TITLE
Nub of rendering challenge

### ISSUE THREAD

**Issue by shaunc** [shaunc/ember-grid#2]
Given

      {{#eg-body}}
        {{my-component value="name"}}
      {{/eg-body}}

We need to render

        {{my-component value="name"}}

In the context of the individual row. So:

1) look up contents of "eg-body" for given column
2) render contents in a custom wrapper component

What is relevant "contents of 'eg-body'? The dom nodes aren't sufficient. What we really want is the template (template string)? The question is, is that available?

**Comment by BryanCrotaz** [comment]
Yes that's the nub I was heading towards but you found the detail. Hmm...

**Comment by BryanCrotaz** [comment]
Wonder if we can yield it in somehow. Set the render context to off screen then yield

**Comment by shaunc** [comment]
In theory, layout of `eg-body` looks like

  {{yield}}
  
anyway. The question will be, does instance of eg-body have a "template" attribute which we can simply assign to a dynamically created component whose purpose is to pass in the context to template.

**Comment by BryanCrotaz** [comment]
Isn't this exactly what #each has to do?

**Comment by BryanCrotaz** [comment]
Yes! It has layout property. See eg-body component def

**Comment by shaunc** [comment]
#each has a static template... we have to do the same thing but dynamically. So can we do:

    {{#ember-list as |column|}}
       {{display-cell tempate=column.cellTemplate}}
    {{/ember-list}}

**Comment by shaunc** [comment]
`layout` is for the wrapper, `template` is for the contents of the wrapper.

**Comment by shaunc** [comment]
So task 1 is to write a little test w/ 

{{#x-foo}}
  something
{{/x-foo}}

{{#ember-list as |foo|}}
  {{render-it template=foo.template}}
{{#ember-list}}

... [will try it out -- now I'm being called away from the keyboard :)]

**Comment by shaunc** [comment]
hmm... so in a block component, `template` is null, and `layout` contains an already compiled htmlbars structure, which itself just contains the "yield". Of course, the DOM is present, but to get back to the actual template we would need to reverse compile.

What we really want (or at least what I was thinking about) is an htmlbars block helper that would grab the unrendered template callback, and save it for us to call back once we have set up the context.

However, the interfaces to this were called private and unstable. I think that (e.g.) in `eg-body` we actually have to iterate over the data and pass it in. If necessary we can use [ember-wormhole](https://github.com/yapplabs/ember-wormhole) to rearrange the content. I'll tweak the readme slightly and try to get an example of this working.

**Comment by BryanCrotaz** [comment]
Yes we definitely need it before rendering. How does #each do it? Isn't that doing the same thing, repeatedly rendering the same template with different injected data?

**Comment by BryanCrotaz** [comment]
Gotcha.

Can't use block helpers any more so we have to use wormhole technique. Can't use wormhole as it requires an elementId. So I've used the same technique in `eg-render/eg-header-cell` and `eg-render/eg-footer-cell`. Works beautifully.

Default header is `column.header`. Default footer is `column.footer`.

Now we just need to get the body working. I suggest you do that as you know `ember-collection` better than me. I'll get header and footer rows looking nice and getting column resize to work.

**Comment by shaunc** [comment]
#each -- I looked at the code and is pretty simple. It wouldn't be hard to copy. The problem is whenever they change the private APIs and #each breaks, they have to fix it. But they don't have to fix ember-grid.

I think we can use wormhole: we can create an id from row*column. And I think we have to to keep the current form, as if we keep the bodies inside the columns, they won't render in the right order. 

This is my plan:

1) Render data fields in columns (with display = none)

2) Render a sheet of divs with ids

Both of the above with ember-collection.

3) Teleport the data fields into place with wormhole.
I naively believe that if a row*column doesn't exist in the destination grid, the wormhole will simply fail without bad side effects.

**Comment by shaunc** [comment]
If you have a better idea tell me.. possibly we could leave as columns and render the whole as a column table, but it won't be what users would expect at least.

**Comment by BryanCrotaz** [comment]
#each can't work. We can't do our own block helpers any more.

#wormhole can't work as we need unique ids - what happens if the user wants two tables?

I've used the same technique as wormhole, and made it work nicely for headers and footers (including fallback to default string)

**Comment by BryanCrotaz** [comment]
take a look at the currently checked in code and see what you think.

**Comment by shaunc** [comment]
A random string would do as a prefix ... but checking out the code now.

**Comment by shaunc** [comment]
Ok .... yes that is good for headers and footers. We can move things around "relatively" rather than absolutely using parent. ... but that isn't the hard part with the body (unique ids per se aren't a problem): here body cells can render a column of fields, but we can't leave them like that -- for instance we need a scrollbar for whole sheet rather than one for each column, so we have to undertake more transformative surgery.

I didn't get as much done as I expected as we have guests, but will consider both techniques. Right now I'm just getting to the point where I have two structures: one with fields in columns, and another with an empty grid of rows & columns. I'll rebase based on your work before I continue --  no need for teleportation when normal transportation already works. Another part I haven't quite worked out yet is sequencing: I can't move elements into the grid before it exists.

**Comment by shaunc** [comment]
Its gotten late and I didn't get done as much as I hoped as a few other things came up. I haven't merged body work yet. Just pushed version with your components in pod structure so we don't diverge too much. Tomorrow have "real work" until evening; hopefully can push something then. Thanks!

**Comment by BryanCrotaz** [comment]
I think we should find a way to use ember container for the body. It's
huge and we shouldn't rewrite it.

**Comment by shaunc** [comment]
ember-collection, you mean?

**Comment by shaunc** [comment]
Ok -- have pushed a version -- shows body rendering. It is still quite buggy: I and ember-collection are both fiddling with contents dynamically and the interference shows. But illustrates the basic technique.

**Comment by BryanCrotaz** [comment]
I've cleaned up jshint and simplified some of the structure.
Wondering if we can eliminate `ember.run.next` to speed up rendering

**Comment by BryanCrotaz** [comment]
I'll see if I can get drag-resize headers next

**Comment by shaunc** [comment]
Great -- I'll check it out. I've found a few bugs in ember-collection which might or might not help on scroll. (Still fixing now....) One more complex thing thing we need to get working is horizontal scroll if the columns push beyond the viewport. What is your npm username so I can add you as npm owner?

**Comment by BryanCrotaz** [comment]
bcrotaz

**Comment by shaunc** [comment]
(added in npm)

**Comment by BryanCrotaz** [comment]
horiz scroll needs to be on all three zones, so `ember-collection` and hread and footer row should have `min-width = sum(column widths) + borders`

**Comment by shaunc** [comment]
yes... I was also thinking of using ember-collection in definition with horizontal window so that we don't have to draw all the columns but only those that fit.

As for borders, I'm wondering if we can get away with leaving that to the user - just render the cells absolutely with no gaps. and let them style eg-cell-body (for example) and and/or whatever else the have actually rendered.

**Comment by BryanCrotaz** [comment]
no, css is king here. we can query css params to see how much space to leave.  Take a look at ember-split-view in the minSize computed property.
Then the user can just use css to style and we use whatever they used.

**Comment by shaunc** [comment]
yes -- very good. Btw perhaps its time to start adding tests? Seems like we have "proof of concept".

**Comment by BryanCrotaz** [comment]
I think we need to prototype a couple more things first.
It's very slow to start up - all the run.next cycles - can we get involved in the rendering so it renders correctly first time?

**Comment by shaunc** [comment]
One thing I'm going to try now is to eliminate the per-column ember-collections body definitions -- no need to use frame animation inside of display==None. Instead we should just render just a slice. ...but its already getting to have non-trivial complexity so lets at least try not to add too many more features before working on the tests.

**Comment by BryanCrotaz** [comment]
That's weird - the divs in the collection are in reverse order/

Not at all sure what your latest comment means

**Comment by BryanCrotaz** [comment]
made some nice progress. We need to think about how body cells are laid out - I think your method is about to hit a brick wall when we try and have the column know about styling to calc offsets. What's wrong with `inline-block`?

**Comment by shaunc** [comment]
merging changes -- about to check in (so you can see what it means).

I thought about `inline-block` ... but if we really want to use ember-collection, it places everything absolutely and would be quite difficult to get around.

wrt to:

    classNames: ['eg-body-cell'],

Can we make sure to namespace our css (it doesn't have to be `eg-body-cell` -- could be `eg-cell` for instance if you want.

**Comment by shaunc** [comment]
wrt to ` the divs in the collection are in reverse order/` -- there is a scrolling bug now. Problem is that ember-collection keeps a hash of elements -- not necessarily in order and then places them and tells them to render as needed. However, render for me means "take your content from the definition" and now if another cell has already taken it I'm stuck with the old content. I need to create another map.

**Comment by shaunc** [comment]
wrt
    classNames: ['cell']

nevermind... I guess thats ok.

**Comment by BryanCrotaz** [comment]
My thinking is that we're namespaced within ember-grid and by using less we enforce that

**Comment by BryanCrotaz** [comment]
Also wondering whether the grid should have two drawing modes. One for a normal table where columns are all visible and one spreadsheet style where a subset of cols are visible. Former doesn't need an ember-collection per row

**Comment by shaunc** [comment]
Yes -- I decided that it was ok. The only thing is if user drops in a table inside something styled `#my-div .cell { ... }` but is easy to fix and OTOH convenient to use. Just I currently use selector to look up cell, and will add `eg-body-cell` as well.

WRT to css styling of cells. how do you want to do it? It seems to me that `column.width` has to govern here...?

**Comment by BryanCrotaz** [comment]
You can look up cell with '.ember-grid .body .cell' or you can start find within body or row for speed

**Comment by BryanCrotaz** [comment]
I've already set cell width from column width

**Comment by shaunc** [comment]
> Former doesn't need an ember-collection per row

The idea would 2 ember collections: one large one which ranges over whole including definitions and governs horizontal styling, and one internal one over rows which governs vertical scrolling. However, we could leave out the large one possibly -- lets get the vertical scrolling working and then consider.

> I've already set cell width from column width

what do you see as the problem w/ absolute positioning?

**Comment by BryanCrotaz** [comment]
I've bodged it quickly with a style binding but need to do it properly later to stay within CSP rules

**Comment by shaunc** [comment]
ok -- but that isn't a problem with absolute position per se, right -- we need to do it right however? (btw you might have noticed I tossed in "unsafe-self" in dev to get rid of annoying reports)

**Comment by BryanCrotaz** [comment]
The problem with absolute positioning is to bring in CSS padding margins and borders. CSS might be overridden in app. The way to do that is in the cell component. Maybe col width is just the content width of the column. Then the cell component adds on CSS widths

**Comment by shaunc** [comment]
I thought of "eg-body-cell" as a wrapper around the "real cell" which is whatever the user wants it to be (and would be best target of css).

**Comment by shaunc** [comment]
we could explicitly render a wrapper and also a cell perhaps, and use the cell for positioning?

**Comment by BryanCrotaz** [comment]
Given the number of cells we ought to minimise dom complexity. And just do what is necessary.

**Comment by BryanCrotaz** [comment]
I wonder if we should start with body rows using the same technique as header rows

**Comment by shaunc** [comment]
In my branch here I've replaced the "hidden" ember-collections with simple slices (which helps with speed)... (no longer working after merge). Thus the body rows from the definition are in order. However, the ones from ember-collection are still an unordered map. Do you have something concrete in mind by " body rows using the same technique as header rows"?

**Comment by shaunc** [comment]
btw -- does getting rid of `run.next` in "willInsertElement" actually work for you? I tried and got lots of warnings about modifying properties during render cycle.

**Comment by BryanCrotaz** [comment]
I think it worked for header and footer but not body. I don't understand the glimmer life cycle yet. 

By same as header, I mean a simple each...header-cell with inline-block positioning. 

Worth speed testing?

**Comment by shaunc** [comment]
I'll give inline-block a whirl ... let me get merged and at least push a branch so you can see what I'm doing at least. After that I want to dive in to this annoying scrolling problem. For me it already renders faster without the hidden ember-collections in the definitions.

**Comment by BryanCrotaz** [comment]
Aha!

**Comment by BryanCrotaz** [comment]
I'll start doing the draggable column boundaries

**Comment by BryanCrotaz** [comment]
If we can get what we have plus resize columns then we just need tests for a first release

**Comment by shaunc** [comment]
:)

**Comment by shaunc** [comment]
ok -- pushed; now on to scrolling

**Comment by BryanCrotaz** [comment]
default headers are broken. investigating

**Comment by BryanCrotaz** [comment]
column resize worked first time. The architecture seems to be working.

You can mark a column as `resizable: false`

**Comment by BryanCrotaz** [comment]
age is marked as not resizable in the dummy app

**Comment by shaunc** [comment]
excellent -- pulled and had fund resizing. How is rendering speed for you now?

**Comment by BryanCrotaz** [comment]
Much faster. But row ordering is completely broken and some rows have no data in them. Startup speed is slow (about 1 second on my machine to fully render)

**Comment by BryanCrotaz** [comment]
Just got booked to do a talk on this component on 10th Sept at Ember London. There's some really interesting techniques in here.

**Comment by shaunc** [comment]
Wow -- great! 

I'm working on ordering now. VS startup speed perhaps we should try turning off rendering of display table until definitions have been processed.

**Comment by shaunc** [comment]
Ordering I should be able to get working but won't be ideal. For further improvement make want to derive from ember-collection so we can take advantage of the same map it is using.

**Comment by BryanCrotaz** [comment]
hmm. Can we not have a `body-row` component which deals with laying out the row so that `ember-collection` only deals with row ordering?

**Comment by shaunc** [comment]
I have added and then deleted a `body-row` three times. :) ... it doesn't really help the situation. Currently `ember-collection` is already only dealing with row ordering. Just that it expects rows to "re-render" when it remaps, and as for me rerendering means taking DOM notes off the shelf from the definition, I also have to put them back on the shelf when I'm done with them -- that is what I'm working on now.

**Comment by BryanCrotaz** [comment]
aha - they were dissappearing down the worm-plug-hole?

**Comment by shaunc** [comment]
Yes -- and so the old contents is still there out of order.

**Comment by BryanCrotaz** [comment]
I've implemented min and max column widths both in column def and in css. I'm going to make styling behave properly re CSP and then I think rendering rows is the only thing left

**Comment by shaunc** [comment]
Speed-wise -- we'll see how it looks. It could be further improved if (a subclass of) ember-collection moved the nodes itself rather than asking for a rerender.

**Comment by BryanCrotaz** [comment]
I wonder if having `body-row` makes your problem neater if not easier as it's a clean place to put the wormhole behaviour

**Comment by BryanCrotaz** [comment]
I think slightly slow if fine for now - let's build tests once we have a v1 functionality. I'm amazed at how fast this has come together. Liking component-style building

**Comment by shaunc** [comment]
right now I've got it pretty well encapsulated: 

The definition cell is responsible for getting node:

    getCellElement(rowIndex) {
      var element = this.element;
      if (element == null || element.childNodes == null) {
        return null;
      }
      var {'_body.offset': offset, '_body.limit': limit } = this.getProperties(
        '_body.offset', '_body.limit');
      if (rowIndex < offset || rowIndex >= offset + limit) { return null; }
      return element.getElementsByClassName('eg-body-cell')[rowIndex - offset];
    },

The rendering cell asks for the contents and moves into place

  didInsertElement: function() {
      this._super.apply(this, arguments);
      var source = this.get('column._zones.body.source');
      var sourceElement = source.getCellElement(this.get('rowIndex'));
      this.moveChildren(sourceElement, this.element);
    },
    moveChildren: function (source, target) {
      if (source == null || target == null) { return; }
      while (source.childNodes.length > 0) {
        target.appendChild(source.childNodes[0]);
      }
    }

**Comment by BryanCrotaz** [comment]
blue header is defined in dummy app styling to confirm that app can override styling

**Comment by BryanCrotaz** [comment]
feels a bit dodgy that the definition cell is dealing with offsets and class names. shouldn't it just grab its entire block?

**Comment by shaunc** [comment]
I'm sorry -- thats not the definition cell. Thats the definition body. Your right that I could break it into two steps -- definition body gets cell component, then cell component gets contents. But currently there is no cell component and as this would be the only logic in it its not so bad I think but can clean up if you like. However, first I have to finish up the "put the nodes back" code.

**Comment by BryanCrotaz** [comment]
if you look at how I've done the header, 
```
parentView.set('_column._zones.header.element', this.get('element'));
```
I'm just grabbing the whole element in the definition

Then `eg-render/eg-header-cell` moves its children:

```
  renderHeader: function() {
  	var header = this.get('_header');
  	if (header)
  	{
	    var sourceElement = header.element;
	    var destinationElement = this.get('element');
	    var node = sourceElement.firstChild;
	    var lastNode = sourceElement.lastChild;
	    while(node) {
	      destinationElement.insertBefore(node, null);
	      node = node !== lastNode ? lastNode.parentNode.firstChild : null;
	    }
	  }
  },
```

**Comment by BryanCrotaz** [comment]
probably
```
moveChildren: function (source, target) {
if (source == null || target == null) { return; }
while (source.childNodes.length > 0) {
target.appendChild(source.childNodes[0]);
}
```
should be in a mixin as the header and footer could reuse the same code

**Comment by shaunc** [comment]
(We would need to add a (say) eg-body/eg-cell ... marginal if you ask me)

**Comment by shaunc** [comment]
mixin: +1

**Comment by BryanCrotaz** [comment]
ok, let's review when you're done

**Comment by shaunc** [comment]
ok -- there you go. header looks good. Its still a bit too lazy -- scroll lags behind. Simplest way to fix would probably to draw the table a few rows larger than it was then put sentinels at top and bottom.

**Comment by shaunc** [comment]
I'm also very surprised it works so well already. One ceveat is its pointing at my private copy of ember-collection, which has some bug fixes still sitting in a PR. Still very good -- I think you've done a great job filling out the architecture. 

Take a look at what I've done. I can put in eg-render/body-row and/or  eg-body/eg-cell if you think its worthwhile.

**Comment by shaunc** [comment]
btw -- should we remove eg-render components from app/ as they are not part of the public interface?

**Comment by shaunc** [comment]
one small thing with header drag: how to resize last cell

**Comment by BryanCrotaz** [comment]
I tried removing eg-render components, but then they're not found by the resolver. Would be nice to find a way around that.

I forgot about the last cell! Do you think the last cell should fill remaining space? That would make sense in a normal table view. min-width settings would work to make the next column collapse when it reached its limit.

Header resize also lags behind a little - much more than in ember-split-view. I'm not too concerned right now.

**Comment by BryanCrotaz** [comment]
list-view had functionality for providing a scroll buffer either side of the visible window. Does ember-collection not have that?

**Comment by BryanCrotaz** [comment]
scroll is broken with your new code. First page renders correctly, then nothing else as I scroll down

![image](https://cloud.githubusercontent.com/assets/2016523/9342014/727d318e-45f1-11e5-9e2f-c7d34ba2aa29.png)

**Comment by shaunc** [comment]
hmm.. not what I see... perhaps  -- can you pull the latest shaunc/ember-collection?

**Comment by BryanCrotaz** [comment]
can you change package.json to point to the correct branch?

**Comment by BryanCrotaz** [comment]
ah! I see - you've fixed ember-collection since I last npm installed.
npm updating...

**Comment by shaunc** [comment]
buffer: it keeps a buffer of old nodes, but tells to rerender just as approaches screen. Arguably a bug.

**Comment by BryanCrotaz** [comment]
and our rerender is slow?

**Comment by shaunc** [comment]
I think its rather that ember executes the hooks with a lag

**Comment by BryanCrotaz** [comment]
so it's already on screen when the hook kicks in?

**Comment by shaunc** [comment]
yes -- exactly. Is it working for you now? (My son awaits my reading to him :)) Just pushed a small fix to give cells height.

**Comment by BryanCrotaz** [comment]
scrolling slowly is fine after update, but fast scroll gives lots of empty rows

**Comment by BryanCrotaz** [comment]
go and read!

**Comment by shaunc** [comment]
:) -- leave any ideas -- back in an hour

**Comment by shaunc** [comment]
(I may be able to fix this in ember-collection)

**Comment by BryanCrotaz** [comment]
I would prefer that rowheight (& anything non interactive) was done in css. Way more efficient, and user can easily override.

**Comment by shaunc** [comment]
hmm... well ember-collection needs rowheight to know how the rows fill out the space... will consider

**Comment by BryanCrotaz** [comment]
we can read it from css? Look at minwidth of header cell to see how. Otherwise it's a grid property. No real problem with it being that, just css is nicer if possible.

Currently body cells are 1 px too narrow from col 2 onwards. Can't see why.

**Comment by shaunc** [comment]
rowHeight -- to read from css we'll need a row to style. Maybe I can put an option in ember-collection to pass in a class name for the item wrapper.

**Comment by BryanCrotaz** [comment]
Ah yes of course. Ok leave as is then

**Comment by BryanCrotaz** [comment]
added alternate row colours and horiz align for default column layouts.

`eg-body` and `eg-render/eg-body-cell` differ in structure from equivalent header and footer. See which you think is best of the two. We should consolidate the design so it's the same everywhere.

**Comment by BryanCrotaz** [comment]
do we still need `column.offset`?

**Comment by shaunc** [comment]
I’ll take a look — busy today but have time in eve.

**Comment by shaunc** [comment]
I have committed a buffering fix to shaunc/ember-collection -- if you pull it should scroll better; we can tweak a little to improve this further now that buffering is working.  Haven't looked at our code yet today but will consider the architecture questions you brought up later on this eve.

**Comment by BryanCrotaz** [comment]
:( Rendering still broken on scroll. On slow scroll it's fine, but scrolling fast results in empty rows.

**Comment by shaunc** [comment]
:( for me too -- though I've been fooling with the parameters. The other issue is if you jerk the scroll around you can easily get doubles of cells rendered at an offset. I'm beginning to worry that the `requestAnimationFrame` isn't (easily) compatible with our swapping out domain nodes. Still have a few more ideas but we may have to take closer control -- e.g. insuring that during scroll elements are rendered before they seen.

**Comment by shaunc** [comment]
update -- even if I get rid of the slice in eg-body and render all the data (which hurts startup and isn't realistic for large datasets) its better but still not perfect. I'll check in this mode in a branch. If you think its ok, we might be able to approach this performance by filling out the buffer incrementally in the background. 

If not, though, we'll have to take more direct control of rendering, perhaps explicitly instantiating elements in the buffer rather than letting animation simply call "rerender" for async rerender as its now done.

**Comment by shaunc** [comment]
I've push something into the "slow-load" branch.

In it I've completely disabled slicing the original content rendering (hence the "slow-load"). I wanted to see how much of the problem is unavoidable with ember-collection, as we are ways to optimize the slicing.

On the front page in the slow-load branch is our table and below it a raw "ember-collection" table for comparison. You can see the ember-collection table also has flicker; with ours its certainly worse.

In console you can see log as it renders. If you watch with a big scroll,, it does a bunch of rows, then waits then another bunch.... Even if the rows are long offscreen, they are still gumming up the queue.

If we are ok with the following, doing approximately the same with incremental slices of content wouldn't be too hard. What do you think?

I am leaning toward the thought that we may have to take over from ember-collection. If we could respond to scroll directly and move a whole chunk of rows at once (keeping in order) we'd be a lot better off than the current randomly ordered & absolutely placed, where even if you are already offscreen when you get control its hard to figure out what to do instead.

**Comment by shaunc** [comment]
@BryanCrotaz My plan is currently to derive from ember-collection and override the `updateCells` method. Instead of just deciding which cells will display, and allowing rerender to lazily take care of it, I will greedily get the content from the body buffer.

[Do you know how to include one addon from another](https://gist.github.com/kristianmandrup/ae3174217f68a6a51ed5#gistcomment-1555556)? I am having trouble including the component so that I can extend from it.

**Comment by BryanCrotaz** [comment]
when scroll happens, what do you get given in the args?

Don't know how to include an addon, but I'm about to add in a mixin from an addon so I'll hit the same problem

**Comment by shaunc** [comment]
Currently, when scroll happens, ember-collection marks the "cells" property as changed, which triggers re-render. In "update-cells", ember reassigns which cells display what (from a map), updates their properties and re-renders them. The cell is an ember-collection place-holder, with underlying row & index as data elements, as well as style info for abs position ec uses it to render the block content:

    {{~#each cells as |cell|~}}
      <div style={{{cell.style}}}>{{#unless cell.hidden}}{{yield cell.item cell.index }}{{/unless}}</div>
    {{~/each~}}

So the row gets the data and index as args. However, the rerender is lazy and in fast scroll it will engage in rendering elements that have already scrolled by.

Instead I'd like to override `updateCells` to check the actual scrollTop currently, and copy the data in for the displayed slice immediately without waiting for re-render.

(I just had one more thought, though, before i do this -- perhaps I can get a PR into ec that at least reupdates offset given current scrollTop when the async rerender happens. May help to be only one async rerender away from current scroll rather than current two -- first for rerender of EC and 2nd for rerender of content cells.)

**Comment by BryanCrotaz** [comment]
these issues do seem like generic ember-collection issues rather than specific to us

**Comment by BryanCrotaz** [comment]
our part should be limited to deciding what to render in each row

**Comment by shaunc** [comment]
They are difficult to deal with generically: generically, asking subcomponent to rerender is the best you can do. A few possible optimizations I could try to get into ember-collection: 

* Possibly you could also anticipate scroll by looking at several datapoints and try to calculate velocity, but that would be tricky to do well.

* You could keep cells in an ordered ring rather than in a map. In this case, when we get a rerender in the row, we could decide -- "you asked me to rerender for row X but I know that row X has already scrolled off, and my position in the ring means I should really render row Y"

However, the former is tricky and bug-prone (and would be hard to get into official repo I think). The latter is really cheating anyway.

I will soon push another copy of my own EC -- see what you think. This will probably be the best we can do with limiting ourselves to 

> deciding what to render in each row

**Comment by BryanCrotaz** [comment]
isn't the latter saying:
1. child cell is being reused so kick off a lazy rerender
2. when rerender comes around cell might have been repurposed so render the correct contents

Your update sounds about right. in slow scroll you're updating off screen rows. In fast scroll you're updating on screen rows so lazy is not important.

**Comment by shaunc** [comment]
> UPDATE and both of these feel like generic ember-collection issues

Ember collection will have a hard time doing better though. In our case if we know we need an immediate update we could do a low-level grab from the body-def buffers. Generically, an async rerender is only possibility.

My plan for this now is to create a new component -- "ember-scroll-ring" which uses the layout-bin-packer to allow virtual scroll of a fixed ring of divs, and allows user to register a callback to do something to an element when it appears or disappears. Then we (in this project) can concentrate on rerendering rows as you say, but we'd be able to do it synchronously with scroll (modulo having to render the body defs in the background).

However, I am currently going to leave this, and go on to cleanup, testing, and then data-driven columns.

**Comment by BryanCrotaz** [comment]
cleanup - have you had a chance to compare your cell declaration/rendering with mine and choose which is best?

I'm looking at the best way to achieve CSP friendly styling - two addons to choose from which have a mixing to do it.

I have a dev here who's finishing up some low level Raspberry Pi stuff, then I'll put him onto writing some tests for this.

What do you think of the doc style I've started in the readme? Could you add docs for the features of the body cells that you've created (eg rowheight)?

**Comment by shaunc** [comment]
> cleanup - have you had a chance to compare your cell declaration/rendering with mine and choose which is best?

I like yours. Will rearrange.

> I'm looking at the best way to achieve CSP friendly styling - two addons to choose from which have a mixing to do it.

Good
> I have a dev here who's finishing up some low level Raspberry Pi stuff, then I'll put him onto writing some tests for this.

Excellent! :)

> What do you think of the doc style I've started in the readme? Could you add docs for the features of the body cells that you've created (eg rowheight)?

I like the style in general. The "incremental buildup" is nice at the start. It may get complex later though. 

I will want to put in a column definition iterator that iterates over columns provided as data once I cleanup my stuff. We will probably want to add other stuff as well.

Perhaps e.g. where we have "add a custom footer" we drop the code for custom header?

**Comment by BryanCrotaz** [comment]
> The "incremental buildup"

It's not strictly observed - I've stripped out some earlier features as I demo the next ones

> columns provided as data
Not sure how we'll do the block content when providing columns as data...

> Perhaps e.g. where we have "add a custom footer" we drop the code for custom header?
Don't know what you mean here

**Comment by shaunc** [comment]
| Perhaps e.g. where we have "add a custom footer" we drop the code for custom header?

In your custom footer example, the example also includes a custom header. I didn't notice the earlier drops; I was suggesting we drop the custom header in the custom footer example.

**Comment by BryanCrotaz** [comment]
>  I was suggesting we drop the custom header in the custom footer example.

I deliberately put both in so that it doesn't look as though you can only have one or the other

**Comment by shaunc** [comment]
I've put in documentation for body and for overall window. (Some of it is aspirational -- have ability to call accessor in my branch; the ability to look up `width`, `height and `rowWidth` are still not present.)

**Comment by shaunc** [comment]
For the moment to get going I'm in the process of writing one test per component... a little bit more than the default "it renders" but not much.

**Comment by shaunc** [comment]
vs ember-collection, we have to land https://github.com/emberjs/ember-collection/issues/26 in some form before we can stop using my branch of ember-collection. Do you want me to merge in those commits you proposed? They look like they are on the fast track anyway.

I am expanding the demo app to include a gallery showing different templates containing ember grids and their results.

+1 for restructure to fully use pods

**Comment by BryanCrotaz** [comment]
I had intended my pr to go against your repo but couldn't work out how

**Comment by BryanCrotaz** [comment]
No idea why tests are failing. I'm getting different results on first and subsequent runs!

```
Integration | Component | eg render/eg body cell: it renders (4, 0, 4)Rerun38 ms
Uncaught Error: you must set a resolver with `testResolver.set(resolver)`@ 6 ms
Source: 	
http://localhost:7357/assets/vendor.js:39469
Died on test #2     at Object.test (http://localhost:7357/assets/test-support.js:1947:11)
    at http://localhost:7357/assets/dummy.js:1221:15
    at mod.state (http://localhost:7357/assets/vendor.js:150:29)
    at tryFinally (http://localhost:7357/assets/vendor.js:30:14)
    at requireModule (http://localhost:7357/assets/vendor.js:148:5)
    at Object.TestLoader.require (http://localhost:7357/assets/test-loader.js:29:9)
    at Object.TestLoader.loadModules (http://localhost:7357/assets/test-loader.js:21:18): Cannot read property 'set' of undefined@ 19 ms
Source: 	
TypeError: Cannot read property 'set' of undefined
    at http://localhost:7357/assets/dummy.js:1226:9
    at Object.wrapper (http://localhost:7357/assets/test-support.js:1929:29)
    at Object.Test.run (http://localhost:7357/assets/test-support.js:3696:28)
    at http://localhost:7357/assets/test-support.js:3825:11
    at process (http://localhost:7357/assets/test-support.js:3384:24)
    at begin (http://localhost:7357/assets/test-support.js:3429:2)
    at http://localhost:7357/assets/test-support.js:3445:4
Uncaught TypeError: Cannot read property 'subject' of undefined@ 24 ms
Source: 	
http://localhost:7357/assets/vendor.js:39469
Expected 2 assertions, but 3 were run@ 38 ms
Source: 	
    at Object.test (http://localhost:7357/assets/test-support.js:1947:11)
    at http://localhost:7357/assets/dummy.js:1221:15
    at mod.state (http://localhost:7357/assets/vendor.js:150:29)
    at tryFinally (http://localhost:7357/assets/vendor.js:30:14)
    at requireModule (http://localhost:7357/assets/vendor.js:148:5)
    at Object.TestLoader.require (http://localhost:7357/assets/test-loader.js:29:9)
    at Object.TestLoader.loadModules (http://localhost:7357/assets/test-loader.js:21:18)

```

**Comment by shaunc** [comment]
vis ember-collection -- will get into my repo

vis tests -- I stepped back for a bit from tests. Apparently the new ember-cli release this week will resolve some problems with ember-cli-mocha and wanted to delay final decision about using it until that hits. In the meantime I have been filling out some examples in "gallery" in dummy app that won't replace tests, but will illustrate features in conjunction with documentation (and have already led me to a few bugs).
